Add an indexedDB file storage example: image-reference-collector

rpl opened the pull request:
This PR contains the first webextension example related to storing files in the IndexedDB storage available to the WebExtensions.

- [x] example built with babel and webpack
- [x] example without any build step
- [x] remove React from the "example without any build step"
- [x] apply wbamberg review comments

wbamberg commented:
@rpl , I will take a look at this as soon as I've got the Firefox 55 doc updates done :).

wbamberg commented on store-collected-images/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## What it does
+
+This example adds a context menu which target any image element in the webpage,
```
target->targets

wbamberg commented on store-collected-images/README.md:
```diff
@@ -0,0 +1,52 @@
+when the context menu action is clicked, the extension open a window and add
+the related image element to the preview list of the collected images,
+the user can then store the collected images by giving the collection a name
+and pressing the **save** button.
```
So, I'm horribly picky because of being a writer, but: this sentence is too long. Better:

This example adds a context menu which targets any image element in the webpage. When the context menu item is clicked, the add-on opens a window and adds the related image element to the preview list of the collected images. The user can then store the collected images by giving the collection a name
and pressing the **save** button.

wbamberg commented on store-collected-images/README.md:
```diff
@@ -0,0 +1,52 @@
+the user can then store the collected images by giving the collection a name
+and pressing the **save** button.
+
+Once a collection of reference images has been stored by the extension, they
```
Again, picky, but I try to use "add-on" in the docs, not "extension".

wbamberg commented on store-collected-images/README.md:
```diff
@@ -0,0 +1,52 @@
+[yarn](https://yarnpkg.com/):
+
+    npm install
+
```
I think at this point it would also be good to document npm run build in case someone doesn't want to use web-ext:

To build the extension, type:

    npm run build

wbamberg commented on store-collected-images/README.md:
```diff
@@ -0,0 +1,52 @@
+
+    npm install
+
+Now you can build the sources and start the extension in one command:
```
"You can also build the sources and start the extension in one command:"

wbamberg commented on store-collected-images/README.md:
```diff
@@ -0,0 +1,52 @@
+
+    npm run start
+
+This creates the source bundles for the WebExtension in the `extension` subdirectory.
```
If you do the change suggested above, this sentence should also be moved or possibly just removed.

wbamberg commented on store-collected-images/README.md:
```diff
@@ -0,0 +1,52 @@
+
+While this npm script is running, any time you edit a file, it will be rebuilt automatically).
+
+[![entension demo screencast](screenshots/screenshot.png "extension demo screencast")](https://youtu.be/t6aVqMMe2Rc)
```
entension->extension (or add-on :-)

wbamberg commented on store-collected-images/src/navigate-collection.js:
```diff
@@ -0,0 +1,67 @@
+"use strict";
+
+import {getFileStorage} from 'idb-file-storage/src/idb-file-storage';
```
I'm not sure you need this import? Isn't it only used by `./utils/image-store`?

wbamberg commented on store-collected-images/src/popup.js:
```diff
@@ -0,0 +1,112 @@
+"use strict";
+
+import {getFileStorage} from 'idb-file-storage/src/idb-file-storage';
```
Again, not sure you need this.

wbamberg commented on store-collected-images/src/popup.js:
```diff
@@ -0,0 +1,112 @@
+          collectedBlobs: [],
+        });
+
+        setTimeout(() => {
```
I'm not sure what these setTimeouts are doing, but I'll take your word for it :-).

wbamberg commented on store-collected-images/src/utils/image-store.js:
```diff
@@ -0,0 +1,36 @@
+"use strict";
```
Since this is really the key function (AIUI) in terms of illustrating the usage of the storage API, it would be great to have some comments.

rpl commented:
Hi @wbamberg I've applied the review comments (and also rebased this PR on top of the eslint PR and applied the needed fixes, eslint still complains of some unused vars, which are actually exported and used in the other scripts but eslint is not recognizing the "exported comment") and I have rewrote the example without React and webpack (and as we agreed in SF, both the versions are available in the PR under different sub directories).

Let me know how it looks to you (the "plain webextension" version).

wbamberg commented:
Looks great @rpl . Thanks for all your work on this.
